db/sqlinterfaces.c

```diff
@@ -7755,16 +7755,36 @@ static int process_set_commands(struct sqlclntstate *clnt)
                     rc = ii + 1;
                 }
             } else if (strncasecmp(sqlstr, "user", 4) == 0) {
+                char *savptr = sqlstr;
                 sqlstr += 4;
                 sqlstr = cdb2_skipws(sqlstr);
-                clnt->have_user = 1;
-                strncpy(clnt->user, sqlstr, sizeof(clnt->user));
+                if (strlen(sqlstr) >= sizeof(clnt->user)) {
+                    sqlstr = savptr;
+                    logmsg(LOGMSG_USER,
+                           "set user parameter can not exceed length of %d "
+                           "characters\n", sizeof(clnt->user) - 1);
+                    rc = ii + 1;
+                } else {
+                    clnt->have_user = 1;
+                    strncpy(clnt->user, sqlstr, sizeof(clnt->user) - 1);
+                    clnt->user[sizeof(clnt->user) - 1] = '\0';
+                }
             } else if (strncasecmp(sqlstr, "password", 8) == 0) {
+                char *savptr = sqlstr;
                 sqlstr += 8;
                 sqlstr = cdb2_skipws(sqlstr);
-                clnt->have_password = 1;
-                strncpy(clnt->password, sqlstr, sizeof(clnt->password));
-                sqlite3Dequote(clnt->password);
+                if (strlen(sqlstr) >= sizeof(clnt->password)) {
+                    sqlstr = savptr;
+                    logmsg(LOGMSG_USER,
+                           "set password parameter can not exceed length of %d "
+                           "characters\n", sizeof(clnt->password) - 1);
+                    rc = ii + 1;
+                } else {
+                    clnt->have_password = 1;
+                    strncpy(clnt->password, sqlstr, sizeof(clnt->password) - 1);
+                    clnt->password[sizeof(clnt->password) - 1] = '\0';
+                    sqlite3Dequote(clnt->password);
```

Comment:
I think the password should be dequoted first before validating its length. Ditto for username.

Reply:
good point.

Reply:
Good idea. Will apply.